File: app/MMarket/c/src/data/models/Qmob.c
// Copyright 24-Jan-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "data/models/Qmob.h"
#include <stddef.h>
#include "dmc/ADouble.h"
#include "dmc/AInt.h"
#include "data/modelDocs/qmobDoc.h"

static void calc (
  AADouble *closes,
  double *params,
  void (*action)(ADouble *closes, ADouble *refs)
) {
  double gap = params[0];

  ADouble **pcloses = closes->es;
  ADouble *refs = aDouble_new();
  AInt *is_solds = aInt_new();


  ADouble *row = *pcloses;
  double *prow = row->es;
  while (prow < row->end) {
    double q = *prow++;
    aDouble_push(refs, q * (1 - gap));
    aInt_push(is_solds, 0);
  }

  while (pcloses < closes->end) {
    ADouble *row = *pcloses++;
    ADouble *new_refs = aDouble_new();
    AInt *new_is_solds = aInt_new();

    double *prow = row->es;
    double *prefs = refs->es;
    int *pis_solds = is_solds->es;
    while (prow < row->end) {
      double q = *prow++;
      double ref = *prefs++;
      int is_sold = *pis_solds++;

      if (is_sold) {
        if (q > ref) {
          aInt_push(new_is_solds, 0);
          aDouble_push(new_refs, q * (1 - gap));
        } else {
          aInt_push(new_is_solds, 1);
          double new_ref = q * (1 + gap);
          aDouble_push(new_refs, new_ref < ref ? new_ref : ref);
        }
      } else {
        if (q < ref) {
          aInt_push(new_is_solds, 1);
          aDouble_push(new_refs, q * (1 + gap));
        } else {
          aInt_push(new_is_solds, 0);
          double new_ref = q * (1 - gap);
          aDouble_push(new_refs, new_ref > ref ? new_ref : ref);
        }
      }
    }

    refs = new_refs;
    is_solds = new_is_solds;

    action(row, refs);
  }
}

Model *qmob_new (void) {
  return model_new(
    "QMOV",
    "Quantum móvil",
    qmobDoc_get(),
    achar_new_from("Intervalo", NULL),
    aDouble_new_c(1, (double[]){0.10})->es,
    aDouble_new_c(1, (double[]){0.005})->es,
    aDouble_new_c(1, (double[]){0.001})->es,
    calc
  );
}

File: app/KtMMarket/go/data/model/qmob.go
// Copyright 04-Jul-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// Model Mobil Quantum.
package model

import (
	"github.com/dedeme/KtMMarket/data/model/docs/qmobDoc"
)

func qmobCalc(
	closes [][]float64,
	params []float64,
	action func(closes, refs []float64),
) {
	gap := params[0]
	nCos := len(closes[0])
	pvrow := make([]float64, nCos)
	for i := 0; i < nCos; i++ {
		for _, cs := range closes {
			if cs[i] > 0 {
				pvrow[i] = cs[i]
				break
			}
		}
		if pvrow[i] == 0 {
			pvrow[i] = 1
		}
	}

	isSolds := make([]bool, nCos)
	refs := make([]float64, nCos)
	for i, c := range pvrow {
		refs[i] = c * (1 - gap)
	}

	for _, cs := range closes {
		newRefs := make([]float64, nCos)
		newIsSolds := make([]bool, nCos)

		for i, c := range cs {
			if c < 0 {
				newRefs[i] = refs[i]
				newIsSolds[i] = isSolds[i]
				continue
			}

			rf := refs[i]
			isSold := isSolds[i]

			if isSold {
				if c > rf {
					// newIsSolds[i] = false
					newRefs[i] = c * (1 - gap)
				} else {
					newIsSolds[i] = true
					newRef := c * (1 + gap)
					if newRef > rf {
						newRef = rf
					}
					newRefs[i] = newRef
				}
			} else {
				if c < rf {
					newIsSolds[i] = true
					newRefs[i] = c * (1 + gap)
				} else {
					// newIsSolds[i] = false
					newRef := c * (1 - gap)
					if newRef < rf {
						newRef = rf
					}
					newRefs[i] = newRef
				}
			}

		}
		refs = newRefs
		isSolds = newIsSolds

		action(cs, refs)
	}

}

func newQmob() *T {
	return &T{
		"QMOV",
		"Quantum móvil",
		qmobDoc.Get(),
		[]string{
			"Intervalo",
		},
		[]float64{
			0.01,
		},
		[]float64{
			0.005,
		},
		[]float64{
			0.001,
		},
		qmobCalc,
	}
}
